### PR TITLE
Convert requirements.txt to pyproject.toml

### PR DESCRIPTION
**Pull request by julienp**
# Description

When using a template that uses requirements.txt with the poetry toolchain, we convert this to a pyproject.toml automatically.

Fixes https://github.com/pulumi/pulumi/issues/16355
Ref https://github.com/pulumi/pulumi/issues/15937
Stacked on top of https://github.com/pulumi/pulumi/pull/16347

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [x] I have formatted my code using `gofumpt`

### CODE REVIEW

File: sdk/python/toolchain/poetry.go [hunk 1]
```diff
@@ -40,6 +43,19 @@ func newPoetry(directory string) (*poetry, error) {
 func (p *poetry) InstallDependencies(ctx context.Context,
 	root string, showOutput bool, infoWriter, errorWriter io.Writer,
 ) error {
+	// If pyproject.toml does not exist, but we have a  requirements.txt,
```
Comment (lunaris) [inline sdk/python/toolchain/poetry.go:46]:
```suggestion
	// If pyproject.toml does not exist, but we have a requirements.txt,
```

File: sdk/python/toolchain/poetry.go [hunk 2]
```diff
@@ -40,6 +43,19 @@ func newPoetry(directory string) (*poetry, error) {
...
+	// generate a new pyproject.toml.
+	pyprojectToml := filepath.Join(root, "pyproject.toml")
+	if _, err := os.Stat(pyprojectToml); err != nil && errors.Is(err, os.ErrNotExist) {
+		requirementsTxt := filepath.Join(root, "requirements.txt")
+		if _, err := os.Stat(requirementsTxt); err != nil && errors.Is(err, os.ErrNotExist) {
+			return fmt.Errorf("could not find pyproject.toml in %s", root)
```
Comment (lunaris) [inline sdk/python/toolchain/poetry.go:52]: Shouldn't the error reference `requirements.txt`?

File: sdk/python/toolchain/poetry.go [hunk 3]
```diff
@@ -151,3 +167,107 @@ func (p *poetry) virtualenvPath(ctx context.Context) (string, error) {
...
+func (p *poetry) convertRequirementsTxt(requirementsTxt, pyprojectToml string) error {
+	f, err := os.Open(requirementsTxt)
+	if err != nil {
+		return fmt.Errorf("failed to open %q", requirementsTxt)
+	}
+	defer f.Close()
+
+	deps, err := dependenciesFromRequirementsTxt(f)
+	if err != nil {
+		return fmt.Errorf("failed gather dependencies from %q", requirementsTxt)
```
Comment (lunaris) [inline sdk/python/toolchain/poetry.go:180]:
```suggestion
		return fmt.Errorf("failed to gather dependencies from %q", requirementsTxt)
```

**Review by lunaris — APPROVED** [review]

**Comment by Frassle** [comment]
Can we add a test for this functionality?

File: changelog/pending/20240610--sdk-python--automatically-converts-requirements-txt-to-pyproject-toml-when-using-poetry.yaml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+changes:
+- type: feat
+  scope: sdk/python
+  description: Automatically converts requirements.txt to pyproject.toml when using Poetry
```
Comment (tgummerer) [inline changelog/pending/20240610--sdk-python--automatically-converts-requirements-txt-to-pyproject-toml-when-using-poetry.yaml:4]:
```suggestion
  description: Automatically convert requirements.txt to pyproject.toml when using Poetry
```

Comment (julienp) [reply changelog/pending/20240610--sdk-python--automatically-converts-requirements-txt-to-pyproject-toml-when-using-poetry.yaml]: Oops, thanks!

**Comment by julienp** [comment]
> Can we add an integration test for this functionality?

Oops, I should have done that! I added one here https://github.com/pulumi/pulumi/pull/16354/commits/ce221341e29136ed6ed4e314e33a2fb6efabc34f

File: sdk/python/toolchain/poetry.go [hunk 5]
```diff
@@ -151,3 +167,107 @@ func (p *poetry) virtualenvPath(ctx context.Context) (string, error) {
...
+
+	w := &bytes.Buffer{}
+	if err := toml.NewEncoder(w).Encode(pp); err != nil {
+		return "", err
+	}
+	return w.String(), nil
+}
+
+func dependenciesFromRequirementsTxt(r io.Reader) (map[string]string, error) {
+	versionRe := regexp.MustCompile("[<>=]+.+")
```
Comment (justinvp) [inline sdk/python/toolchain/poetry.go:237]: Will this work for operators like `~=`?

https://peps.python.org/pep-0508/#grammar

File: sdk/python/toolchain/poetry.go [hunk 6]
```diff
@@ -151,3 +167,107 @@ func (p *poetry) virtualenvPath(ctx context.Context) (string, error) {
 	}
 	return virtualenvPath, nil
 }
+
+func (p *poetry) convertRequirementsTxt(requirementsTxt, pyprojectToml string) error {
```
[the remark this answers is not in the document]
Comment (julienp) [reply sdk/python/toolchain/poetry.go:171]: No, neither does the `~=` syntax work. I limited the support at what we use in our templates.

There are also things like `httpx[http2]` which translates to
`httpx = {version = "0.27.0", extras = ["http2"]}`. And then there are various URL schemes you could have to download the package/source from somewhere.

**Pull request merged**